### PR TITLE
feat(remote-config): realtime config updates

### PR DESCRIPTION
**Pull request by Lyokone**
### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [x] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [x] `jest` tests added or updated in `packages/\*\*/__tests__`
- [x] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java [hunk 1]
```diff
@@ -26,13 +26,17 @@
...
 import com.google.firebase.remoteconfig.FirebaseRemoteConfigInfo;
 import com.google.firebase.remoteconfig.FirebaseRemoteConfigSettings;
 import com.google.firebase.remoteconfig.FirebaseRemoteConfigValue;
+import com.google.firebase.remoteconfig.FirebaseRemoteConfigException;
+import com.google.firebase.remoteconfig.FirebaseRemoteConfigFetchThrottledException;
 import io.invertase.firebase.common.UniversalFirebaseModule;
-import java.util.HashMap;
-import java.util.Map;
+import com.facebook.react.bridge.*;
+import java.util.*;
```
Comment (Lyokone) [inline packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java:39]: Shouldn't we import each Class individually?

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java [hunk 2]
```diff
@@ -53,12 +57,12 @@ Task<Void> fetch(String appName, long expirationDuration) {
...
-          FirebaseRemoteConfig config = FirebaseRemoteConfig.getInstance(firebaseApp);
-          Tasks.await(expirationDuration == -1 ? config.fetch() : config.fetch(expirationDuration));
-          return null;
-        });
+      getExecutor(),
+      () -> {
+        FirebaseRemoteConfig config = FirebaseRemoteConfig.getInstance(firebaseApp);
+        Tasks.await(expirationDuration == -1 ? config.fetch() : config.fetch(expirationDuration));
+        return null;
+      });
```
Comment (Lyokone) [inline packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java:65]: Are you using the same formatter as the project?

File: packages/remote-config/e2e/config.modular.e2e.js [hunk 3]
```diff
@@ -743,5 +749,14 @@ describe('remoteConfig() modular', function () {
         should(supported).equal(true);
       });
     });
+
+    describe('onConfigUpdated', function () {
+      it('onConfigUpdated can run without an issue', async function () {
+        const { getRemoteConfig, onConfigUpdated } = remoteConfigModular;
+        const remoteConfig = getRemoteConfig();
+
+        onConfigUpdated(remoteConfig, function (event, error) {});
```
Comment (Lyokone) [inline packages/remote-config/e2e/config.modular.e2e.js:758]: I think this test need to be completed

File: packages/remote-config/ios/RNFBConfig/RNFBConfigModule.m [hunk 4]
```diff
@@ -223,6 +223,31 @@ + (BOOL)requiresMainQueueSetup {
...
+        }
+
+        callback(@[
+            @{
+                @"updatedKeys" : [configUpdate.updatedKeys allObjects],
+            },
+            userInfo != nil ? userInfo : [NSNull null],
+        ]);
+    }];
+}
```
Comment (Lyokone) [inline packages/remote-config/ios/RNFBConfig/RNFBConfigModule.m:249]: Is there any way to remove the listener once Javascript has stopped listening?

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java [hunk 5]
```diff
@@ -143,9 +147,53 @@ Task<FirebaseRemoteConfigInfo> ensureInitialized(String appName) {
     return ensureInitializedTask;
   }
 
+  void onConfigUpdated(String appName, Callback callback) {
+    FirebaseApp firebaseApp = FirebaseApp.getInstance(appName);
+
+    FirebaseRemoteConfig.getInstance(firebaseApp).addOnConfigUpdateListener(new ConfigUpdateListener() {
```
Comment (russellwheatley) [inline packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java:153]: This event listener should probably be tracked like we do with auth [here](https://github.com/invertase/react-native-firebase/blob/main/packages/auth/android/src/main/java/io/invertase/firebase/auth/ReactNativeFirebaseAuthModule.java#L96) and removed using [this overridden method](https://github.com/invertase/react-native-firebase/blob/main/packages/auth/android/src/main/java/io/invertase/firebase/auth/ReactNativeFirebaseAuthModule.java#L117). This means you will have to move it to the `ReactNativeFirebaseConfigModule`.  Remote Config has multi-app support so you will need to track them on a per Firebase app basis so the implementation shouldn't be too far removed from how we implement the authStateListener [here](https://github.com/invertase/react-native-firebase/blob/main/packages/auth/android/src/main/java/io/invertase/firebase/auth/ReactNativeFirebaseAuthModule.java#L152-L181). i.e. use the app name as a key for the listener.

File: packages/remote-config/ios/RNFBConfig/RNFBConfigModule.m [hunk 6]
```diff
@@ -223,6 +223,31 @@ + (BOOL)requiresMainQueueSetup {
   }
 }
 
+RCT_EXPORT_METHOD(onConfigUpdated
+                  : (FIRApp *)firebaseApp
+                  : (RCTResponseSenderBlock)callback) {
+    [[FIRRemoteConfig remoteConfigWithApp:firebaseApp] addOnConfigUpdateListener:^( FIRRemoteConfigUpdate * _Nonnull configUpdate, NSError * _Nullable error) {
```
Comment (russellwheatley) [inline packages/remote-config/ios/RNFBConfig/RNFBConfigModule.m:229]: as mentioned above for android will also be required for iOS.

**Review by russellwheatley — CHANGES_REQUESTED** [review]

File: packages/remote-config/e2e/config.modular.e2e.js (hunk 3 above)
Comment (exaby73) [reply packages/remote-config/e2e/config.modular.e2e.js:758]: I couldn't think of a test here. Followed tests/integration_test/firebase_remote_config/firebase_remote_config_e2e_test.dart in this case. In the comment there, you mentioned that it's not possible to change values on the fly to test the callback. Any ideas on how to improve this test?

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java (hunk 2 above)
Comment (exaby73) [reply packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java:65]: I'm using IntelliJ, and it formatted the code like this (force of habit to hit the reformat code keybind). I'm not familiar with Java formatters. Could you point me to a command I can run to fix this formatting or should I manually revert the formatting changes?

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java (hunk 1 above)
Comment (exaby73) [reply packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java:39]: Done :)

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java (hunk 2 above)
Comment (russellwheatley) [reply packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java:73]: That's a good question actually, I can't see a script in the `package.json` for formatting android. What is the correct way @mikehardy?

File: packages/remote-config/android/src/reactnative/java/io/invertase/firebase/config/ReactNativeFirebaseConfigModule.java [hunk 7]
```diff
@@ -17,23 +17,54 @@
...
 
+  @Override
+  public void onCatalystInstanceDestroy() {
+    super.onCatalystInstanceDestroy();
+
+    Iterator configRegistrationsIterator = mConfigUpdateRegistrations.entrySet().iterator();
+
+    while (configRegistrationsIterator.hasNext()) {
+      Map.Entry pair = (Map.Entry) configRegistrationsIterator.next();
+      String appName = (String) pair.getKey();
```
Comment (russellwheatley) [inline packages/remote-config/android/src/reactnative/java/io/invertase/firebase/config/ReactNativeFirebaseConfigModule.java:60]: this isn't used. Can remove.

File: packages/remote-config/ios/RNFBConfig/RNFBConfigModule.m [hunk 8]
```diff
@@ -223,6 +247,41 @@ + (BOOL)requiresMainQueueSetup {
...
+            [userInfo setValue:@"unknown" forKey:@"code"];
+            [userInfo setValue:error.localizedDescription forKey:@"message"];
+            [userInfo setValue:@(error.code) forKey:@"nativeErrorCode"];
+            [userInfo setValue:error.localizedDescription forKey:@"nativeErrorMessage"];
+        }
+
+        callback(@[
+            @{
+                @"updatedKeys" : [configUpdate.updatedKeys allObjects],
+            },
```
Comment (russellwheatley) [inline packages/remote-config/ios/RNFBConfig/RNFBConfigModule.m:269]: if there is an error, I'm not sure we should be returning the first value, it should probably be `[NSNull null]`.

File: packages/remote-config/android/src/reactnative/java/io/invertase/firebase/config/ReactNativeFirebaseConfigModule.java (hunk 7 above)
Comment (exaby73) [reply packages/remote-config/android/src/reactnative/java/io/invertase/firebase/config/ReactNativeFirebaseConfigModule.java:60]: This code is to remove all listeners on destroy, right? This is how auth is doing it

File: packages/remote-config/ios/RNFBConfig/RNFBConfigModule.m [hunk 9]
```diff
@@ -223,6 +247,41 @@ + (BOOL)requiresMainQueueSetup {
...
+                  : (FIRApp *)firebaseApp
+                  : (RCTResponseSenderBlock)callback) {
+    FIRConfigUpdateListenerRegistration *newRegistration = [[FIRRemoteConfig remoteConfigWithApp:firebaseApp] addOnConfigUpdateListener:^( FIRRemoteConfigUpdate * _Nonnull configUpdate, NSError * _Nullable error) {
+        NSMutableDictionary *userInfo;
+
+        if (error != nil) {
+            userInfo = [NSMutableDictionary dictionary];
+
+            [userInfo setValue:@(NO) forKey:@"fatal"];
+            [userInfo setValue:@"unknown" forKey:@"code"];
```
Comment (russellwheatley) [inline packages/remote-config/ios/RNFBConfig/RNFBConfigModule.m:260]: We should probably try to propagate the original error code `error.code` before making it defacto `unknown`.

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java [hunk 10]
```diff
@@ -143,9 +155,53 @@ Task<FirebaseRemoteConfigInfo> ensureInitialized(String appName) {
...
+        results.put("updatedKeys", updatedKeysList);
+        callback.invoke(Arguments.makeNativeMap(results), null);
+      }
+
+      @Override
+      public void onError(FirebaseRemoteConfigException error) {
+        WritableMap userInfoMap = Arguments.createMap();
+
+        if (error == null) {
+          userInfoMap.putString("code", "unknown");
```
Comment (russellwheatley) [inline packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java:177]: same deal here, try to find the error code first before making it `unknown`.

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java [hunk 11]
```diff
@@ -143,9 +155,53 @@ Task<FirebaseRemoteConfigInfo> ensureInitialized(String appName) {
...
+        if (error.getCause() instanceof FirebaseRemoteConfigFetchThrottledException) {
+          userInfoMap.putString("code", "throttled");
+          userInfoMap.putString(
+            "message",
+            "fetch() operation cannot be completed successfully, due to throttling.");
+        } else {
+          userInfoMap.putString("code", "failure");
+          userInfoMap.putString(
+            "message",
+            "fetch() operation cannot be completed successfully.");
```
Comment (russellwheatley) [inline packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java:194]: use `error.getMessage()` here rather than hardcoded value like you do with iOS.

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java [hunk 12]
```diff
@@ -143,9 +155,53 @@ Task<FirebaseRemoteConfigInfo> ensureInitialized(String appName) {
...
+        Map<String, Object> results = new HashMap<>();
+        results.put("updatedKeys", updatedKeysList);
+        callback.invoke(Arguments.makeNativeMap(results), null);
+      }
+
+      @Override
+      public void onError(FirebaseRemoteConfigException error) {
+        WritableMap userInfoMap = Arguments.createMap();
+
+        if (error == null) {
```
Comment (russellwheatley) [inline packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java:176]: I don't think this can be `null`. Remove.

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java [hunk 13]
```diff
@@ -143,9 +155,53 @@ Task<FirebaseRemoteConfigInfo> ensureInitialized(String appName) {
...
+          userInfoMap.putString("code", "throttled");
+          userInfoMap.putString(
+            "message",
+            "fetch() operation cannot be completed successfully, due to throttling.");
+        } else {
+          userInfoMap.putString("code", "failure");
+          userInfoMap.putString(
+            "message",
+            "fetch() operation cannot be completed successfully.");
+        }
```
Comment (russellwheatley) [inline packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java:195]: I think remove all this and propagate error as it is. Here are the docs for `FirebaseRemoteConfigException`. https://firebase.google.com/docs/reference/kotlin/com/google/firebase/remoteconfig/FirebaseRemoteConfigException

You need to use `@NonNull` annotations on this parameter like this:
```
public void onError(@NonNull FirebaseRemoteConfigException error) {
```
Are you using android studio for android dev? It will flag when you need to use it.

File: packages/remote-config/ios/RNFBConfig/RNFBConfigModule.m (hunk 8 above)
Comment (exaby73) [reply packages/remote-config/ios/RNFBConfig/RNFBConfigModule.m:269]: Fixed this

**Review by russellwheatley — CHANGES_REQUESTED** [review]
some updates required but looking good 😎

File: packages/remote-config/android/src/reactnative/java/io/invertase/firebase/config/ReactNativeFirebaseConfigModule.java (hunk 7 above)
Comment (russellwheatley) [reply packages/remote-config/android/src/reactnative/java/io/invertase/firebase/config/ReactNativeFirebaseConfigModule.java:60]: where is appName used?

Comment (russellwheatley) [reply packages/remote-config/android/src/reactnative/java/io/invertase/firebase/config/ReactNativeFirebaseConfigModule.java:60]: If you use android studio, it will tell you when variables aren't used:
Screenshot 2023-05-18 at 11 14 44

Comment (exaby73) [reply packages/remote-config/android/src/reactnative/java/io/invertase/firebase/config/ReactNativeFirebaseConfigModule.java:60]: Ahh sorry didn't see the appName thingy. Removed it :)

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java (hunk 12 above)
Comment (exaby73) [reply packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java:176]: Done

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java (hunk 2 above)
Comment (mikehardy) [reply packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java:74]: We (that is, myself on behalf of Invertase) repackage https://github.com/google/google-java-format in https://www.npmjs.com/package/google-java-format so we have autoformatting, and it is available here: https://github.com/invertase/react-native-firebase/blob/a58ea4bfd72c903d43844da30d4f3bf5feb0f057/package.json#L11

If you run that command locally until it finally succeeds (it may take quite a few runs, it is a bit unreliable unfortunately but it will always eventually run to completion) then all your files will be formatted locally to our standard

File: packages/remote-config/e2e/config.modular.e2e.js (hunk 3 above)
Comment (mikehardy) [reply packages/remote-config/e2e/config.modular.e2e.js]: I think the tests are inadequate and I'm working on a test style that uses a cloud function I will deploy (I'm working on this currently, nearly ready).

The cloud function will have access to the admin SDK which allows it to read/write the remote-config template, and will have an API that lets you add a remote config template key, remove a remote config template key, and update a remote config template key value as a list of 0 or more operations that it will do at once.

So then in the e2e tests I expect this to happen:

- single listener addition/removal case:
  - we test updating a known remote-config template key value while there are no listeners, no listeners are called
  - we add a listener, do same update, listener should be called
  - we remove the listener, do same update, no listener should be called

- update type verification
  - make sure key addition is handled correctly
  - same with key removal
  - same with key update of all update types

- react-native reload verification
  - add a listener, verify it is called when you update the remote-config template
  - do a react-native hot reload, make sure the listener is removed down to native level

- multiple listener addition/removal accounting verification
  - add one listener, make sure it is called when template updates
  - add second listener, make sure both are called when template updates
  - remove first listener, make sure only second listener is called
  - add third/fourth listener, make sure 2nd/3rd/4th are called
  - remove 3rd listener, make sure 2nd/4th are called
  - remove all listeners, make sure none are called


All of this is predicated on the ability to trigger the listener, which is based on the assumption we'll have a function+admin-sdk deployed on the project as an enabler - once that's up, the ability to e2e test this thing should be pretty thorough

File: packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java (hunk 2 above)
Comment (exaby73) [reply packages/remote-config/android/src/main/java/io/invertase/firebase/config/UniversalFirebaseConfigModule.java]: Following @mikehardy's advice, I ran the formatter. It did change a few files, but these I think are now in sync with our code style

File: packages/remote-config/lib/index.js [hunk 14]
```diff
@@ -282,6 +289,45 @@ class FirebaseConfigModule extends FirebaseModule {
...
+          Promise.resolve().then(() => {
+            listener(event, undefined);
+          });
+          return;
+        }
+
+        Promise.resolve().then(() => {
+          listener(undefined, event);
+        });
+      },
```
Comment (russellwheatley) [inline packages/remote-config/lib/index.js:316]: I think this can be removed now, this is specific to auth logic

**Comment by TaylorDale** [comment]
Have been looking for this... thanks for the work 😎

File: tests/app.js [hunk 15]
```diff
@@ -40,9 +40,9 @@ import * as remoteConfigModular from '@react-native-firebase/remote-config';
 import '@react-native-firebase/storage';
 import * as storageModular from '@react-native-firebase/storage';
 import jet from 'jet/platform/react-native';
-import React from 'react';
 import { AppRegistry, Button, NativeModules, Text, View } from 'react-native';
 import DeviceInfo from 'react-native-device-info';
+import React from 'react';
```
Comment (mikehardy) [inline tests/app.js:45]: Please avoid unnecessary changes in PRs, this file would not be touched at all except for this change, which is not related to the PR, so it just adds noise to the diff / increases review burden

**Review by mikehardy — COMMENTED** [review]
I've completed implementation of test scaffolding for this feature on branch https://github.com/invertase/react-native-firebase/tree/%40mikehardy/remote-config-realtime-updates

You can use the `updateTemplate` function from `helpers.js` to execute a batch of delete/update/add operations against the existing template, which will enable listener testing to verify this works.

We need to coordinate how to integrate the work, but in this PR I notice a few things:

- there is no update to the index.d.ts file so there are no typescript typings, what exactly is sent to the listener when the config is updated?
- I do not see complete support for multiple listeners? it appears there is a one-to-one limitation between app and listener, but I think you can have more than one listener per app? Is that an intentional constraint on the implementation to limit it to one listener per app?
- I do not see any documentation of the feature? (I have started some on my branch and was expecting a conflict here, but the docs are not updated here...)
- as noted above - I do not see testing

My branch has a sketch in the e2e file of the testing I expect to see to validate the feature

If you would like to continue working on it, you can perhaps rebase this branch on to my branch (just dropping the Podfile.lock commit, most likely...) and then start working through the e2e test file with the idea that your implementation should then be exercised

Follow that up with typescript typings and making sure the docs match the API you've chosen and we should be able to ship

What do you think?

File: tests/app.js (hunk 15 above)
Comment (exaby73) [reply tests/app.js]: Done

File: packages/remote-config/lib/index.js (hunk 14 above)
Comment (exaby73) [reply packages/remote-config/lib/index.js]: Done

**Comment by mikehardy** [comment]
I unpicked the merge commits which made a hash of the commit trail, rebased main back into the branch (vs merge), squashed the remaining changes on this branch into one change (since it is one feature addition) and re-pushed.

Do *not* merge into this branch, rather re-fetch it - `git fetch origin && git reset --hard origin/feat/remote-config-onconfigupdated`

I'm working on now on listener changes and automating the testing of the listener using the cloud function implemented earlier, will re-push when those are stable locally.

**Comment by mikehardy** [comment]
single listener test is more less stable now and definitely tests the back-and-forth of the listener triggering via use of the cloud function to update the config in realtime during the e2e test.

It is yielding some unexpected results that bear examination though - the full set of keys in the remote config are coming back as updated each time at the moment, which does not match with upstream docs

the branch has also been rebased to include #7180 so it is up to date with main

**Comment by mikehardy** [comment]
single listener e2e test now definitely appears to be stable and types documentation is updated in an attempt to explain the strong need to activate the new config after the callback is called in order to get results that make sense

last step is to handle multiple listeners in an efficient manner by de-subscribing the native listener when no more javascript listeners exist. will have to get that tomorrow morning my time (~12 hrs from now) then we can shipit

**Comment by mikehardy** [comment]
This works in e2e now with multiple listeners working and with optimized native subscribe/unsubscribe to make sure we don't open a persistent web socket unless necessary.

However, there is a problem locally for me where my android emulator is no longer receiving realtime updates at all and I need to determine why prior to cleaning up the commit chain and releasing this. Apologies for the delay but it's close!

**Comment by mikehardy** [comment]
I'm getting unexpected behavior from the underlying firebase-ios-sdk with regard to removing listeners and have opened an issue related: https://github.com/firebase/firebase-ios-sdk/issues/11458

Still working through why android is completely failing now

**Comment by mikehardy** [comment]
Need to clean the commit trail up but I believe the final form of the PR is present now, including a workaround for the firebase-ios-sdk item I uncovered

**Comment by mikehardy** [comment]
firebase-ios-sdk appears to have something else going on where it receives a "template was updated" message on it's websocket but then is not successful at fetching the new remote config template. Android does not show this behavior so I have asked upstream for guidance: https://github.com/firebase/firebase-ios-sdk/issues/11462

The feature will appear pretty flaky on iOS until a reliable method to handle it is implemented, if our e2e CI is any indicator.

**Comment by mikehardy** [comment]
commit stream cleaned up, added a trailing commit that turns on SDK logging while investigating ios template fetch failure

tests are clamped to just realtime remoteconfig tests until the above issue is sorted

ready to go other than that
